### ISSUE TITLE
Extend default query construction with custom expression builders

### ISSUE THREAD

**Issue by Danielku15** [OData/WebApi#1629]
Please note that this is not a bug but rather a feature request or RFC. I tried to search for similar request but could not directly find something.

**Background**
I currently have the task to implement an OData API where entities and their properties are dynamically defined via configuration. You can think of it as a ERP-alike system following a entity-attribute-value model. OData already provides the "open types" system but in my case the model is not fully "open". The properties, types and values are strictly defined. It is a bit like a mixture of the ODataUntypedSample and open types. Unfortunately many features are not supported with open types (filtering, sorting etc.) which means I need a different solution. 

_Disclaimer: I personally would like to prevent such a complex/dynamic system on the backend but it is an essential requirement for our data model that administrators can define new models and properties. At least on API layer I would like to establish a clean strict typed data model._

Let's assume I have the dynamic EDM generation covered as this is quite nicely described in the [DynamicEdmModelCreation](https://github.com/OData/ODataSamples/tree/master/WebApiCore/DynamicEdmModelCreation) sample. Also setting up the routes for my dynamic entities should be easily possible via some custom routing implementations. 

The problem I'm fighting with, is providing the IQueryable which will result in correct responses based on the input OData query. The default CLR typed implementation already does all the magic of building a expression tree that can be executed by any IQueryProvider (ultimately by EF Core). But the current implementation requires all EDM properties to be mapped to a raw CLR property on the type. 

**Problem 1: Building the query expression**
It would be good to have a built-in way to the ASP.net Core OData framework where I can inject a fallback handler if the CLR type does not directly contain the properties by name. My base interface would expose some EAV properties where I can "easily" provide an expression how to obtain the value for a property by name. Here some pseudo code to show what I would like to do: 

```csharp
interface IDynamicEntity 
{
   IList<DynamicPropertyValue> DynamicProperties { get;set; }
}
class DynamicPropertyValue 
{
    public string StringValue { get; set; }
    public int IntValue { get; set; }
    public double DoubleValue { get; set; }
    public DynamicProperty Property { get; set; }
}
class DynamicProperty
{
    public DynamicPropertyType Type {get; set;}
    public string Name {get;set;}
}
class MyDynamicEntityPropertyExpressionProvider : IODataPropertyExpressionProvider<IDynamicEntity>
{
  public LambdaExpression GetPropertyAccessLambda(string propertyName, Type expectedPropertyType) 
  { 
      // the following lamdbas would actually be similar to ExpressionHelpers.GetPropertyAccessLambda 
      // but to show how the lambda-body look like I added some raw lambda code. 
      if(expectedPropertyType.Equals(typeof(string)))
      {
        return entity => entity.DynamicProperties.Single(p=>p.Property.Name == propertyName).StringValue;
      }
      else if(expectedPropertyType.Equals(typeof(int)))
      {
        return entity => entity.DynamicProperties.Single(p=>p.Property.Name == propertyName).IntValue;
      }
      else if(expectedPropertyType.Equals(typeof(bool)))
      {
        return entity => entity.DynamicProperties.Single(p=>p.Property.Name == propertyName).IntValue != 0;
      }
      else if(expectedPropertyType.Equals(typeof(double)))
      {
        return entity => entity.DynamicProperties.Single(p=>p.Property.Name == propertyName).DoubleValue;
      }
      // .. other mappings
  }
}
``` 

With such a expression provider in place the existing query infrastructure should be able to build dynamic queries that can be sent to any IQueryProvider (e.g. Entity Framework) and deliver correct results. From what I've seen the existing implementation directly goes down to the `Microsoft.AspNet.OData.ExpressionHelpers.*` methods which build expressions with direct property accesses. If this expression lookup would be pluggable you can build a more dynamic system that covers most EAV and open type scenarios (maybe even DTOs?). I'm not fully aware of all internals and how queries are applied, maybe some more methods are needed to inject all required expressions. But I hope the idea is clear. 

**Problem 2: Returning results**
Maybe this is already covered by the solution of problem 1. After a `IQueryable<T>` is available the result-set also needs to be built correctly. When the OData implementation attempts to load the values for all EDM properties it also needs to query the values via the `IODataPropertyExpressionProvider` if there is no raw CLR property available.

**Conclusion**
With such a solution potentially also some Open Types workflows could be solved where queries currently to not support several options. 

For now the two solutions I see is: Either you go fully untyped and implement the IEdmEntityObject on your own (not sure if queries will work there). Or you clone most of the original OData implementation for applying query options and add your custom extensions. 

The final question: Is such a change to the framework realistic and reasonable?

**Comment by brjohnstmsft** [comment]
@Danielku15 Some teams within Microsoft have implemented OData services with user-defined EDM models, similar to what you're trying to do. I work on one of those teams (Azure Search).

The approach that we use is to avoid `IQueryable` altogether. For result representation, we use `IEdmEntityObject` etc. For queries, we have our own translator that takes the Semantic AST (i.e. `QueryNode`) from `ODataQueryOptions` and translates it into an internal representation that is native to our underlying data store. The filter translator is a large and non-trivial body of code, but this is really the only way I know of to execute OData filters on a "dynamic" model.

Someone from the OData team might have a better proposal, but I thought it would be worth sharing our experience with Azure Search as a data point.

**Comment by Danielku15** [comment]
@brjohnstmsft Thanks a lot for sharing this insight. Your solution sounds a lot like that there is not a lot from the actual OData framework usable in such a dynamic setup (only the query parser and the EDM system?). Developing a whole query translation is quite an overhead for most of the people because it does not only mean to re-implement parts the OData framework, but also potentially the ORM mapping (e.g. Entity Framework in my case). This requires deep knowledge on the different parts and a lot of resources. In cases where you have a very special data store and enough resources available to maintain the custom implementation it might be a feasible solution to re-implement this parts. 

I am currently working on some prototype for our upcoming product. I will definitly give it a try to pull the original OData implementation and patch the required property accesses with an extension system to see if my proposal would work.

**Comment by Danielku15** [comment]
After many investigations on my whole prototype stack I decided to drop the idea of a EAV system. Not only the OData library has serious issues with this kind of structure, also Entity Framework and on DB level there are so many obstacles and issues which lead me to this decision. 

I came partly quite far on realizing such a change to the framework:

* Dynamic EDM generation based on the data model was working
* Querying single-type entity collections were working fine
* Id-querying and $select was working

When I encountered tired to tackle the $filter and $orderby I ran into a [Entity Framework Core issue](https://github.com/aspnet/EntityFrameworkCore/issues/13486) where using a sub-collection entity for further filtering lead to exceptions and also bad query performance. 

With this said I decided to close this issue here. If anybody ever wonders what my solution to this topic was: I will likely go for a Sparse Column system in my database model where I have some spare-columns defined in my model for the dynamic values. On API level I aim to map this generic-multi-purpose columns to strict EDM model properties with clear validations, names etc. 

To achieve this I will likely need open another feature request to request that 1 CLR type can be the data source for multiple EDM types. Internally the Web API OData framework assumes that 1 CLR type only serves 1 EDM type.
